### PR TITLE
Disarm when losing connection

### PR DESCRIPTION
**Pull request by gunturiCM**
## Main Changes

This is an attempt to address issue 29. The issue is to add a fail-safe that shuts down the drone when the radio connection has been lost for a preset amount of period, which is 500ms.

## Associated Issues
- Addresses #29

### CODE REVIEW

**Comment by gunturiCM** [comment]
The code has been tested.

File: include/fail_safes.hpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+       /**
+        * Set the last time in micro-seconds at which radio data was heard
+        * 
+        * This is used to keep track of when the connection might have been lost,
+        * based on which the timeout mechanism triggers.
+        * 
+        * @param t last time in micro-seconds at which the radio was heard
+       */
+       void setLastRadioReceptionTime(long long t);
+
```
Comment (jamie-54) [inline include/fail_safes.hpp:72]: Instead of `t` should it have a meaningful name?

File: platformio.ini [hunk 2]
```diff
@@ -33,7 +33,7 @@ board_build.f_cpu = 240000000L
 ; BZZZ_DEBUG should be used only for debugging
 build_flags =
 	-D BZZZ_DEBUG
-	-D BZZZ_LOGGING_LEVEL=3
+	-D BZZZ_LOGGING_LEVEL=1
```
Comment (jamie-54) [inline platformio.ini:36]: think this should be set back to 3 before merging

File: src/fail_safes.cpp [hunk 3]
```diff
@@ -0,0 +1,56 @@
+#include "fail_safes.hpp"
+#include "util.hpp"
+
+
+namespace bzzz
+{
+    FailSafes::FailSafes(){};
+
+    void FailSafes::setLastRadioReceptionTime(long long t)
```
Comment (jamie-54) [inline src/fail_safes.cpp:9]: Same here, instead of `t` should it have a meaningful name?

File: src/fail_safes.cpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        // Set private variable
+        this->m_lastRadioReceptionTime = t;
```
Comment (jamie-54) [inline src/fail_safes.cpp:12]: and here

File: src/main.cpp [hunk 5]
```diff
@@ -14,8 +15,10 @@ Radio radio;
 AHRS ahrs;
 Controller controller;
 Quaternion initialQuaternion;
+FailSafes failSafes;
 float yawReferenceRad = 0.0;
 float initialAngularVelocity[3];
+float eulerAngles[3];
```
Comment (jamie-54) [inline src/main.cpp:21]: is this being used?

File: src/fail_safes.cpp [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+    bool FailSafes::radioConnectionCheck()
+    {
+        // Check radio connection and timeout if connection had been lost for a preset amount of time.
+        if(micros() - this->m_lastRadioReceptionTime > this->m_radioConnectionTimeoutInMicroseconds)
+        {
+            this->m_HALT_SYSTEM = true;
+            return false;
+        }
+        else
```
Comment (jamie-54) [inline src/fail_safes.cpp:29]: is this else needed?

**Comment by jamie-54** [comment]
@gunturiCM I added a few comments mainly just to tidy up code but it worked well when testing :+1:

File: src/fail_safes.cpp (hunk 6 above)
Comment (alphaville) [reply src/fail_safes.cpp:29]: no

**Review by alphaville — COMMENTED** [review]
Has this been tested on the quadcopter?

**Comment by jamie-54** [comment]
> Has this been tested on the quadcopter?

Yes

File: include/fail_safes.hpp (hunk 1 above)
Comment (gunturiCM) [reply include/fail_safes.hpp:72]: Changed it to `lastRadioReceptionTime`.

File: src/main.cpp [hunk 7]
```diff
@@ -73,10 +84,14 @@ void loop()
   float angularVelocityCorrected[3];
   float controls[3];
 
-  radio.readPiData();
-  if (radio.kill())
+  // if radio data received update the last data read time.
+  if(radio.readPiData()) failSafes.setLastRadioReceptionTime(micros());
```
[the remark this answers is not in the document]
Comment (gunturiCM) [reply src/main.cpp:88]: Done.

File: src/fail_safes.cpp (hunk 6 above)
Comment (gunturiCM) [reply src/fail_safes.cpp:29]: Edited to
```
if(micros() - this->m_lastRadioReceptionTime > this->m_radioConnectionTimeoutInMicroseconds)
{
    this->m_HALT_SYSTEM = true;
    return false;
}
this->m_HALT_SYSTEM = false;
return true;
```

**Comment by gunturiCM** [comment]
Made the changes suggested above.

File: include/fail_safes.hpp [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+namespace bzzz
+{
+    class FailSafes
+    {
+        private:
+        /**
+         * Radio timeout variables for connection loss check
+        */
+       unsigned long long m_lastRadioReceptionTime;
+       unsigned long m_radioConnectionTimeoutInMicroseconds;
```
Comment (alphaville) [inline include/fail_safes.hpp:13]: The type of `m_lastRadioReceptionTime` is `unsigned long long`, but `m_radioConnectionTimeoutInMicroseconds` is `unsigned long`. Is that correct?

File: include/fail_safes.hpp [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+       bool m_HALT_SYSTEM = false;  // Bool to keep keep track if any fail-safe triggers a system halt.
+
+       /**
+        * Poiters to useful objects that are declared outside of this scope
+        * 
+        * Current list of objects pointer:
+        * 1. of class MotorDriver -> this is declared in main.cpp,
+        *                           used to control motors.
+       */
+       MotorDriver *motorDriver;
```
Comment (alphaville) [inline include/fail_safes.hpp:23]: Should be `m_motorDriver`

File: include/fail_safes.hpp [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+        * Currently only stalls the motors,
+        * might need to include a proper shutdown procedure in the future
+       */
+       void shutDown();
+
+       public:
+       /**
+        * Default constructor
+       */
+       FailSafes();
```
Comment (alphaville) [inline include/fail_safes.hpp:49]: I'm going to make this constructor private. One shouldn't be able to create a failsafe object with a null motor driver.

File: include/fail_safes.hpp [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        * 
+        * @param t last time in micro-seconds at which the radio was heard
+       */
+       void setLastRadioReceptionTime(long long lastRadioReceptionTime);
+
+       /**
+        * set the stopMotor function pointer to point to the actual function implementation 
+        * which stalls the motors
+       */
+      void setMotorDriverObjPtr(MotorDriver *motorDriverObj);
```
Comment (alphaville) [inline include/fail_safes.hpp:77]: The motor driver should be provided upon construction. I'm removing this method.

**Review by alphaville — CHANGES_REQUESTED** [review]
@gunturiCM Overall, it looks good. I've made a few changes. Let us test it again on the quadcopter before we merge.

**Review by alphaville — CHANGES_REQUESTED** [review]
Chandra, I made some changes and simplified the API. Now `FailSafes` is independent of `MotorDriver` (see [SOLID principles](https://en.wikipedia.org/wiki/SOLID)).  

I tested using the serial: indeed, if we don't send data for a few ms, it disarms, but we should test on the quadcopter once again.

File: include/fail_safes.hpp (hunk 8 above)
Comment (gunturiCM) [reply include/fail_safes.hpp]: Yes, professor. `m_lastRadioReceptionTime` needs to store up time in microseconds, which we don't know how long. While `m_radioConnectionTimeoutInMicroseconds` is just a positive constant.

**Review by alphaville — APPROVED** [review]
@gunturiCM, if this has been tested we can merge it

**Pull request merged**